File: src/main/java/com/my/admin/app/commerce/service/OrderService.java
package com.my.admin.app.commerce.service;

import com.black.food.manager.dao.OrderDao;
import com.black.food.manager.model.api.request.order.CreateOrderItemsRequest;
import com.black.food.manager.model.api.request.order.CreateOrderRequest;
import com.black.food.manager.model.entity.Order;
import com.black.food.manager.model.entity.OrderItem;
import com.black.food.manager.model.entity.OrderType;
import com.black.food.manager.model.entity.Product;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
public class OrderService {

    private OrderDao orderDao;

    private ProductService productService;

    public Set<OrderItem> createOrderItems(final CreateOrderRequest createOrderRequest) {

        Set<OrderItem> orderItems = new HashSet<>();

        var createOrderItemsRequest = createOrderRequest.getCreateOrderItemsRequest();

        createOrderItemsRequest.forEach(orderItem -> {
            orderItems.add(createOrderItem(orderItem));
        });

        return orderItems;
    }

    private OrderItem createOrderItem(final CreateOrderItemsRequest createOrderItemsRequest) {
        OrderItem orderItem = new OrderItem();

        var product = findProductById(createOrderItemsRequest.getProductId());

        orderItem.setProduct(product);
        orderItem.setCount(createOrderItemsRequest.getCount());

        return orderItem;
    }

    private Product findProductById(final Long productId) {
        return productService.findById(productId);
    }

    public OrderType findOrderTypeById(final Long orderTypeId) {
        return orderDao.findOrderTypeById(orderTypeId);
    }

    public List<Order> findAll() {
        return orderDao.findAll();
    }

    public Set<OrderType> findAllOrderTypes() {
        return orderDao.findAllOrderTypes();
    }

    @Autowired
    public void setOrderDao(final OrderDao orderDao) {
        this.orderDao = orderDao;
    }


    @Autowired
    public void setProductService(final ProductService productService) {
        this.productService = productService;
    }
}

File: src/main/kotlin/com/bk/reactive/app/commerce/admin/my/service/OrderService.kt
package com.bk.reactive.app.commerce.admin.my.service

import com.bk.reactive.app.commerce.admin.my.dao.OrderDAO
import com.bk.reactive.app.commerce.admin.my.dao.ProductDAO
import com.bk.reactive.app.commerce.admin.my.model.api.request.order.CreateOrderItemsRequest
import com.bk.reactive.app.commerce.admin.my.model.api.request.order.CreateOrderRequest
import com.bk.reactive.app.commerce.admin.my.model.document.Order
import com.bk.reactive.app.commerce.admin.my.model.document.OrderItem
import com.bk.reactive.app.commerce.admin.my.model.document.OrderType
import com.bk.reactive.app.commerce.admin.my.model.document.Product
import mu.KotlinLogging
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class OrderService(
    private val orderDAO: OrderDAO,
    private val productDAO: ProductDAO,
    private val exceptionService: ExceptionService
) {
    private val logger = KotlinLogging.logger {}

    fun createOrderItems(createOrderRequest: CreateOrderRequest): Flux<OrderItem> {
        return Flux
            .fromIterable(createOrderRequest.createOrderItemsRequest)
            .flatMap { orderItem -> createOrderItem(orderItem) }
    }

    private fun createOrderItem(createOrderItemsRequest: CreateOrderItemsRequest): Mono<OrderItem> {
        return findProductById(createOrderItemsRequest.productId)
            .map { product ->
                OrderItem(
                    id = null,
                    orderId = null,
                    product = product,
                    count = createOrderItemsRequest.count
                )
            }
    }

    fun findOrderTypeById(orderTypeId: String): Mono<OrderType> {
        return orderDAO.findOrderTypeById(orderTypeId)
            .switchIfEmpty(
                Mono.error {
                    exceptionService.throwNotFoundException("No se encontraron tipos de pedidos.")
                }
            )
    }

    fun findAll(): Flux<Order> = orderDAO.findAll()

    fun findAllOrderTypes(): Flux<OrderType> = orderDAO.findAllOrderTypes()

    private fun findProductById(productId: String): Mono<Product> = productDAO.findById(productId)
}
